### DISCUSSION TITLE
[.NET] Would it be possible to add support for MacCatalyst?

### DISCUSSION THREAD

**Discussion by MartyIX** [jedisct1/libsodium#1235] [Q&A]
I have tried to [create a PR](https://github.com/ektrah/libsodium-core/pull/89) on https://github.com/ektrah/libsodium-core side to add support for Apple MacCatalyst (`net7.0-maccatalyst` runtime identifier in .NET).

Now it [turns out](https://github.com/dotnet/runtime/discussions/79502#discussioncomment-4376529) that the `dylib` for MacCatalyst is not in the nuget package of `jedisct1/libsodium` 

So I wonder whether a modification of 

* https://github.com/jedisct1/libsodium/blob/master/dist-build/macos.sh, and
* https://github.com/jedisct1/libsodium/blob/01c237f1f0dfa2e81f31c5468ccad50bbc582545/packaging/dotnet-core/libsodium.pkgproj#L37

would do the job to add the support for the platform?

I would give a hand but I'm not sure how to add the target `x86_64-apple-ios-macabi` to the makefiles (at least that's what I understand is required to do basically).

What do you think please?

Thanks

Resources:

* https://developer.apple.com/forums/thread/122571?answerId=675312022#675312022 `clang -target x86_64-apple-ios-macabi `
* https://github.com/Instabug/Instabug-iOS/issues/444
* https://stackoverflow.com/questions/70770314/autotools-says-its-building-for-macos-although-i-specified-maccatalyst-as-the

**Comment by jedisct1** [comment]
The `xcframework` includes Catalyst (as well as all iOS, macOS, watchOS and TVOS versions + simulators).

It can be built with `dist-build/apple-xcframework.sh`, or you can simply download pre-built files from [swift-sodium](https://github.com/jedisct1/swift-sodium/tree/master/Clibsodium.xcframework), that are frequently updated.

Since `xcframework` is the standard way to distribute multi-platform Apple libraries, there has to be a way to use them in `.NET` projects.
If not, you can still get files for individual architectures from it.

**Reply by MartyIX** [reply]
Unfortunately, I don't know anything about `xcframework`. But for me to understand better, would be possible to do:

* run `dist-build/apple-xcframework.sh` in `dotnet-core.yml` (similarly to https://github.com/jedisct1/libsodium/blob/01c237f1f0dfa2e81f31c5468ccad50bbc582545/.github/workflows/dotnet-core.yml#L178-L191)
* then pack the final library as ` .libsodium-pack/runtimes/maccatalyst-arm64/native/` (similarly to https://github.com/jedisct1/libsodium/blob/01c237f1f0dfa2e81f31c5468ccad50bbc582545/.github/workflows/dotnet-core.yml#L249-L252)
* and add `<Content Include="runtimes/maccatalyst-arm64/native/libsodium.dylib" PackagePath="runtimes/maccatalyst-arm64/native/" />` to `libsodium.pkgproj` (similarly to https://github.com/jedisct1/libsodium/blob/01c237f1f0dfa2e81f31c5468ccad50bbc582545/packaging/dotnet-core/libsodium.pkgproj#L37)

So it's like: first compile for the maccatalyst platform, then pack the final library in the nuget package.

I might be completely wrong but I would like to understand what would be necessary to make it work.

> Since xcframework is the standard way to distribute multi-platform Apple libraries, there has to be a way to use them in .NET projects.
If not, you can still get files for individual architectures from it. 

So this is really what Rolf [said](https://github.com/dotnet/runtime/discussions/79502#discussioncomment-4376529) yesterday. I verified that in .NET console application on macOS, everything works and that's because your libsodium nuget package contains `osx-arm64` folder with the compiled library. And the only reason why a .NET MacCatalyst application on my macOS does not work with libsodium is because there is no `maccatalyst-arm64` folder in your nuget package (that's my understanding at least).

So I'm just saying that adding compiled native maccatalyst-arm64 (and/or maccatalyst-x64) libsodium library in your libsodium nuget package would solve the issue.

Does it make sense to you?

**Reply by jedisct1** [reply]
What are other Apple targets that should be included?

Looks like .NET also supports iOS and TVOS, so if we pull libraries from the `xcframework`, we should probably include anything .NET supports as well.

What should the directory names be for these different targets? Is there a document listing all the supported targets and the names?

Also, testing these new targets in CI would be useful. Would you mind adding these tests?

**Reply by MartyIX** [reply]
> What are other Apple targets that should be included?

https://github.com/dotnet/designs/blob/main/accepted/2021/net6.0-tfms/net6.0-tfms.md#design mentions `net6.0-ios`, `net6.0-maccatalyst` and `net6.0-tvos`

(I find this description https://github.com/dotnet/runtime/issues/44882#issue-745846137 quite useful)


> What should the directory names be for these different targets?

I believe there should be `maccatalyst-x64`, `maccatalyst-arm64`, `tvos-x64`, `tvos-arm64`, etc.

> Is there a document listing all the supported targets and the names?

There is the [Runtime identifier catalog](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog) and the complete list is [here](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json).


> Also, testing these new targets in CI would be useful. Would you mind adding these tests?

I can try to do something similar to https://github.com/jedisct1/libsodium/blob/c4a9f1b2af809c72e694c98c49563ddef35a7ed2/.github/workflows/dotnet-core.yml#L262-L302. It might take some time, I'm not a Github Actions expert really.

**Reply by ektrah** [reply]
Unfortunately, the story for platforms other than Windows/Linux/macOS isn't as straightforward. For those three, all you have to do is put a pre-compiled `.dll`/`.so`/`.dylib` in the right folder of the NuGet package and the .NET app will load them at runtime using the dynamic linker. However, as I understand it, there's no dynamic linker in iOS, and a pre-compiled `.a` needs to be statically linked into the .NET app at compile-time. The .NET toolchain takes care of this, but this is [severely under-documented](https://github.com/xamarin/xamarin-macios/issues/14647). Adding to the confusion, there is still a lot of information floating around from the Xamarin.iOS days, which I believe is pretty much outdated since .NET 6/7. The story for Android and MacCatalyst might be different; don't know. I'd propose to try to experimentally create a separate NuGet package (not uploaded to nuget.org) that packages the xcframework libraries, create a GitHub Action that compiles a test app for Android/iOS/MacCatalyst demonstrating that the NuGet package works, and then bring the working solution into the libsodium repository.

**Reply by MartyIX** [reply]
Maybe @rolfbjarne could just have a quick look to say what is necessary given his expertise from https://github.com/xamarin/xamarin-macios ?

**Reply by rolfbjarne** [reply]
We support `xcframework`s just fine, although due to how assets are located in nugets, content will end up being duplicated, because:

* An `xcframework` contains bits for all platforms.
* Native assets in nugets go into their own architecture-specific folder (`runtimes/<RuntimeIdentifier>/native`)
* This means that the entire `xcframework` will end up duplicated across all RuntimeIdentifiers (it should be possible to remove content from the `xcframework` that's not applicable to the RuntimeIdentifier in question if size is a big concern).

Note that we also support a zipped version of an xcframework (`*.xcframework.zip`), which is necessary if there are any symlinks in the xcframework, otherwise building from Windows won't produce valid app bundles.

RuntimeIdentifiers for Apple platforms:

* iOS: ios-arm64, iossimulator-arm64, iossimulator-x64
* tvOS: tvos-arm64, tvossimulator-arm64, tvossimulator-x64
* macOS: osx-arm64, osx-x64
* Mac Catalyst: maccatalyst-arm64, maccatalyst-x64

It's correct that for iOS and tvOS, you can't use a `*.dylib` file, the App Store will reject any apps with `*.dylib` files in them. There are two solutions: either use a static library (`*.a`) or a framework (`*.framework`). You can put any of these in an `xcframework`, and we'll link with whatever's there.

It looks like the .xcframework here contains static libraries (`*.a` files), so that should work just fine: https://github.com/jedisct1/swift-sodium/tree/master/Clibsodium.xcframework

**Reply by ektrah** [reply]
Thanks a lot for the explanation, @rolfbjarne!

A few follow-up questions:

* Since all these RIDs import the `unix` RID, would it be an option to place a `Clibsodium.xcframework.zip` in `runtimes/unix/native/` to avoid the duplication?

* Does `macos` still need a `.dylib` or is the `.a` from Clibsodium.xcframework fine? Is there a best practice?

* Is `[DllImport("__Internal")]` still needed or is it now fine to use `[DllImport("libsodium")]` across all platforms?

**Reply by rolfbjarne** [reply]
> * Since all these RIDs import the `unix` RID, would it be an option to place a `Clibsodium.xcframework.zip` in `runtimes/unix/native/` to avoid the duplication?

I'm not entirely sure, but I don't think that will work.

> * Does `macos` still need a `.dylib` or is the `.a` from Clibsodium.xcframework fine? Is there a best practice?

For `macos` either way works, it doesn't really matter.

> * Is `[DllImport("__Internal")]` still needed or is it now fine to use `[DllImport("libsodium")]` across all platforms?

`__Internal`  is required for every platform that uses the static (`*.a`) library.

**Reply by MartyIX** [reply]
Based on the great feedback of @rolfbjarne and @ektrah, I succeeded in making libsodium work in a maccatalyst app on my mac arm64 machine. More specifically, I don't observe `DllNotFoundException` anymore.

Please clone the [sample app](https://github.com/MartyIX/libsodium-maccatalyst-sample) and follow the following instructions if you want to test it too:

```bash
git clone https://github.com/jedisct1/libsodium.git
cd libsodium

# Add the following line to the `packaging/dotnet-core/libsodium.pkgproj` file:
# <Content Include="runtimes/maccatalyst-arm64/native/libsodium.a" PackagePath="runtimes/maccatalyst-arm64/native/" />
#
# Also modify version to "1.0.18.6" (any non-yet-released version is fine for testing but please stick with this one so 
# https://github.com/MartyIX/libsodium-maccatalyst-sample works for you).

# Folder where we're going to pack a .NET NuGet package.
mkdir .libsodium-pack
cp AUTHORS ChangeLog LICENSE packaging/dotnet-core/libsodium.pkgproj .libsodium-pack/

# Compilation requires an Apple machine as far as I know.
dist-build/apple-xcframework.sh

# My machine is an ARM64 one.
mkdir -p .libsodium-pack/runtimes/maccatalyst-arm64/native
cp libsodium-apple/catalyst/lib/libsodium.a .libsodium-pack/runtimes/maccatalyst-arm64/native # You can see that I just copy the static library and then I use `[DllImport("__Internal")]` in the sample app. That's easy for me to understand. 

## If you have x64 mac machine. Use this, I guess.
# mkdir -p .libsodium-pack/runtimes/maccatalyst-x64/native
# cp libsodium-apple/catalyst/lib/libsodium.a .libsodium-pack/runtimes/maccatalyst-x64/native

# Pack the NuGet package.
# The result is located in '.libsodium-pack/bin/Release/libsodium.1.0.18.6.nupkg'
dotnet pack -c Release .libsodium-pack/libsodium.pkgproj

# Copy libsodium.1.0.18.6.nupkg to your https://github.com/MartyIX/libsodium-maccatalyst-sample/tree/main/LocalNugetCache folder.
# Start https://github.com/MartyIX/libsodium-maccatalyst-sample/tree/main/LocalNugetCache using: dotnet build && dotnet run
# You should observe a MacCatalyst app which does not crash and displays the testing output.
```

This is really an attempt to make it work and the goal is to move forward (even if the direction is not correct) so we can get to some final solution. Best effort really.

One notable question is how to test that a MacCatalyst application works because there is no exit code as console applications have. That's maybe something that @rolfbjarne possibly know -- any pointer here would be greatly appreciated.

**Reply by rolfbjarne** [reply]
> One notable question is how to test that a MacCatalyst application works because there is no exit code as console applications have. That's maybe something that @rolfbjarne possibly know -- any pointer here would be greatly appreciated.

You can execute the executable inside the .app directly instead of the app bundle:

```
$ path/to/MyTest.app/Contents/MacOS/MyTest
```

and then you'll get an exit code.

This works for macOS apps as well btw.

**Reply by MartyIX** [reply]
Wow! Exciting. Thanks!

**Comment by jedisct1** [comment]
But can't we remove all the Apple targets we have and ship the `xcframework` file instead?

**Reply by MartyIX** [reply]
That's a good question. I understand Rolf's [comment](https://github.com/jedisct1/libsodium/discussions/1235#discussioncomment-4397018) this way: You can have a single `xcframework` file (the compiled library) but one needs to duplicate it to several folders (i.e. platforms RIDs) in the NuGet package. So the final NuGet file size will be larger than one would like/want. 

@rolfbjarne Is that correct?

**Reply by rolfbjarne** [reply]
> That's a good question. I understand Rolf's [comment](https://github.com/jedisct1/libsodium/discussions/1235#discussioncomment-4397018) this way: You can have a single `xcframework` file (the compiled library) but one needs to duplicate it to several folders (i.e. platforms RIDs) in the NuGet package. So the final NuGet file size will be larger than one would like/want.
> 
> @rolfbjarne Is that correct?

That is correct.

It's also likely possible to shrink each xcframework copy by removing files for other platforms, but that would be a manual and possibly error-prone process.

**Reply by MartyIX** [reply]
With Rolf's big help, I tested the following approach using `libsodium.a` from `libsodium-apple/Clibsodium.xcframework/ios-arm64_x86_64-maccatalyst` folder:

Please clone the [sample app](https://github.com/MartyIX/libsodium-maccatalyst-sample) and follow the following instructions if you want to test it too:

```bash
git clone https://github.com/jedisct1/libsodium.git
cd libsodium

# Add the following line to the `packaging/dotnet-core/libsodium.pkgproj` file:
# <Content Include="runtimes/maccatalyst-x64/native/libsodium.a" PackagePath="runtimes/maccatalyst-x64/native/" />
# <Content Include="runtimes/maccatalyst-arm64/native/libsodium.a" PackagePath="runtimes/maccatalyst-arm64/native/" />
#
# Also modify version to "1.0.18.6" (any non-yet-released version is fine for testing but please stick with this one so 
# https://github.com/MartyIX/libsodium-maccatalyst-sample works for you).

# Folder where we're going to pack a .NET NuGet package.
mkdir .libsodium-pack
cp AUTHORS ChangeLog LICENSE packaging/dotnet-core/libsodium.pkgproj .libsodium-pack/

# Compilation requires an Apple machine as far as I know.
dist-build/apple-xcframework.sh

mkdir -p .libsodium-pack/runtimes/maccatalyst-arm64/native
cp libsodium-apple/Clibsodium.xcframework/ios-arm64_x86_64-maccatalyst/libsodium.a .libsodium-pack/runtimes/maccatalyst-arm64/native

mkdir -p .libsodium-pack/runtimes/maccatalyst-x64/native
cp libsodium-apple/Clibsodium.xcframework/ios-arm64_x86_64-maccatalyst/libsodium.a .libsodium-pack/runtimes/maccatalyst-x64/native

# Pack the NuGet package.
# The result is located in '.libsodium-pack/bin/Release/libsodium.1.0.18.6.nupkg'
dotnet pack -c Release .libsodium-pack/libsodium.pkgproj

# Copy libsodium.1.0.18.6.nupkg to your https://github.com/MartyIX/libsodium-maccatalyst-sample/tree/main/LocalNugetPackages folder.
# Start https://github.com/MartyIX/libsodium-maccatalyst-sample sample app using: `dotnet build && dotnet run`
# You should observe a MacCatalyst app which does not crash and displays the testing output.
```

and it works for me with the [sample app](https://github.com/MartyIX/libsodium-maccatalyst-sample) with arm64 arch and even x64 (emulated on my machine using Rosetta).

**Reply by MartyIX** [reply]
@jedisct1 What is a good next step please?
